### PR TITLE
AAV-164/Graph component tets added

### PR DESCRIPTION
**Pull request by itdev**

### CODE REVIEW

File: src/components/Graph/Graph.spec.js [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+    });
+    expect(graph.instance().updateSelection).toHaveBeenCalled();
+  });
+});
+
+test('shouldComponentUpdate method should return false', () => {
+  const spy = jest.spyOn(Graph.prototype, 'shouldComponentUpdate');
+  const wrapper = shallow(<Graph graph={graphData} />);
+  wrapper.instance().shouldComponentUpdate();
+  expect(spy).toHaveBeenCalled();
```
Comment (Onatolich) [inline src/components/Graph/Graph.spec.js]: This test is not checking what declared in its summary.

File: src/components/Graph/Graph.spec.js [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+test('onEdgeClick should call onEdgeClick prop with edge data', () => {
+  const edge = {
+    source: 'testSource',
+    target: 'testTarget',
+  };
+  const onEdgeClick = jest.fn();
+  const graph = shallow(<Graph graph={graphData} onEdgeClick={onEdgeClick} />);
+  graph.instance().onEdgeClick(edge);
+
+  expect(onEdgeClick).toHaveBeenCalledWith({
```
Comment (Onatolich) [inline src/components/Graph/Graph.spec.js]: `toHaveBeenCalledWith(edge)`

File: src/components/Graph/Graph.spec.js [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+});
+
+describe('refresh method', () => {
+  let graph;
+  let instance;
+  const buffer = {};
+
+  beforeEach(() => {
+    buffer.processGraph = Graph.processGraph;
+    Graph.processGraph = jest.fn().mockReturnValue({ nodes: [], edges: [] });
```
Comment (Onatolich) [inline src/components/Graph/Graph.spec.js]: Use `jest.spyOn()` and `jest.restoreAllMocks()`

File: src/components/Graph/Graph.spec.js [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+      restart: jest.fn(),
+    };
+  });
+
+  afterEach(() => {
+    Graph.processGraph = buffer.processGraph;
+  });
+
+  test('should call restart method', () => {
+    instance.refresh({
```
Comment (Onatolich) [inline src/components/Graph/Graph.spec.js]: Put it in some kind of `const refreshData = {...}` at the beginning of current describe

File: src/components/Graph/Graph.spec.js [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+      graph: {
+        nodes: [],
+        edges: [],
+      },
+    });
+
+    expect(instance.trafficRunner.restart).toHaveBeenCalled();
+  });
+  test('should call applyDiffOnSelection method', () => {
```
Comment (Onatolich) [inline src/components/Graph/Graph.spec.js]: New line

File: src/components/Graph/Graph.spec.js [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+    instance.trafficRunner = {
+      restart: jest.fn(),
+    };
+  });
+
+  afterEach(() => {
+    Graph.processGraph = buffer.processGraph;
+  });
+
+  test('should call restart method', () => {
```
Comment (Onatolich) [inline src/components/Graph/Graph.spec.js]: `should call restart method of trafficRunner`

File: src/components/Graph/Graph.spec.js [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+
+    expect(instance.applyDiffOnSelection).toHaveBeenCalledWith({
+      edges: { add: [], modify: [], remove: [] },
+      nodes: { add: [], modify: [], remove: [] },
+    });
+  });
+});
+
+describe('applyDiffOnSelection method', () => {
+  const graphDiff = {
```
Comment (Onatolich) [inline src/components/Graph/Graph.spec.js]: Move it to upper scope and use it in prev test as well.

File: src/components/Graph/Graph.spec.js [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+    instance.applyDiffOnSelection(graphDiff);
+    expect(props.resetSelection).not.toHaveBeenCalled();
+  });
+
+  test('should not call resetSelection if selectedNode and selectedEdge are not exists', () => {
+    const graph = shallow(<Graph graph={graphData} />);
+    const instance = graph.instance();
+    instance.applyDiffOnSelection(graphDiff);
+    expect(props.resetSelection).not.toHaveBeenCalled();
+  });
```
Comment (Onatolich) [inline src/components/Graph/Graph.spec.js]: There are 2 more conditions when resetSelection will not be called.

File: src/components/Graph/Graph.spec.js [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+  const graph = shallow(<Graph graph={graphData} onEdgeClick={onEdgeClick} />);
+  graph.instance().onEdgeClick(edge);
+
+  expect(onEdgeClick).toHaveBeenCalledWith({
+    source: edge.source,
+    target: edge.target,
+  });
+});
+
+describe('refresh method', () => {
```
Comment (Onatolich) [inline src/components/Graph/Graph.spec.js]: You should also test that new graph is processed by `Graph.processGraph` and set to `graph` attr of instance, `applyDiffOnCollection` is called with every diff entry, `simulation` is called if there are new nodes and `onSimulationEnd` after that.

File: src/components/Graph/Graph.spec.js [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+    graph.instance().refresh = jest.fn();
+    graph.instance().componentWillReceiveProps({
+      graph: {
+        nodes: [],
+        edges: [],
+      },
+    });
+    expect(graph.instance().refresh).toHaveBeenCalled();
+  });
+  test('should call updateSelection if new props selectedNode and selectedEdge are not equal to old ones', () => {
```
Comment (Onatolich) [inline src/components/Graph/Graph.spec.js]: New line

File: src/components/Graph/Graph.spec.js [hunk 11]
```diff
@@ -0,0 +1,222 @@
...
+    graph = shallow(<Graph
+      graph={graphData}
+      selectedNode={selectedNode}
+      selectedEdge={selectedEdge}
+    />);
+    graph.instance().updateSelection = jest.fn();
```
Comment (Onatolich) [inline src/components/Graph/Graph.spec.js]: `const instance = graph.instance();`

File: src/components/Graph/Graph.spec.js (hunk 10 above)
Comment (Onatolich) [reply src/components/Graph/Graph.spec.js]: This test should be splitted by 3 when for changed selectedNode, one for changed selectedEdge and one checking that `updateSelection` is not called if selection is not changed

File: src/components/Graph/Graph.spec.js [hunk 12]
```diff
@@ -0,0 +1,222 @@
...
+describe('componentWillReceiveProps', () => {
+  let graph;
+  let selectedNode;
+  let selectedEdge;
+
+  beforeEach(() => {
+    selectedNode = 1;
+    selectedEdge = { source: 1, target: 2 };
+    graph = shallow(<Graph graph={graphData} />);
+  });
```
Comment (Onatolich) [inline src/components/Graph/Graph.spec.js]: `const instance = graph.instance()` and use it in your tests

File: src/components/Graph/Graph.spec.js [hunk 13]
```diff
@@ -0,0 +1,222 @@
...
+        edges: [],
+      },
+    });
+    expect(graph.instance().refresh).toHaveBeenCalled();
+  });
+  test('should call updateSelection if new props selectedNode and selectedEdge are not equal to old ones', () => {
+    graph = shallow(<Graph
+      graph={graphData}
+      selectedNode={selectedNode}
+      selectedEdge={selectedEdge}
```
Comment (Onatolich) [inline src/components/Graph/Graph.spec.js]: You already have shallowed graph from `beforeEach`

File: src/components/Graph/Graph.spec.js [hunk 14]
```diff
@@ -0,0 +1,222 @@
...
+  let selectedNode;
+  let selectedEdge;
+
+  beforeEach(() => {
+    selectedNode = 1;
+    selectedEdge = { source: 1, target: 2 };
+    graph = shallow(<Graph graph={graphData} />);
+  });
+
+  test('should call refresh if new prop graph are not equal to old one', () => {
```
Comment (Onatolich) [inline src/components/Graph/Graph.spec.js]: Add one more checking that refresh is not called if graph data still the same

File: src/components/Graph/Graph.spec.js [hunk 15]
```diff
@@ -0,0 +1,222 @@
...
+  test('should set true to root element ', () => {
+    expect(root.classed).toHaveBeenCalledWith('simulation', true);
+  });
+
+  test('should set false to root element', () => {
+    graph.instance().simulate();
+    expect(root.classed).toHaveBeenCalledWith('simulation', false);
+  });
+
+  test('should call onSimulationEnd prop with root element', () => {
```
Comment (Onatolich) [inline src/components/Graph/Graph.spec.js]: It's test of `onSimulationEnd` method. Why it's in tests for `componentDidMount`?

File: src/components/Graph/Graph.spec.js [hunk 16]
```diff
@@ -0,0 +1,222 @@
...
+    TrafficRunner.getLoops = buffer.getLoops;
+    TrafficRunner.prototype.tick = buffer.tick;
+  });
+
+  test('should set true to root element ', () => {
+    expect(root.classed).toHaveBeenCalledWith('simulation', true);
+  });
+
+  test('should set false to root element', () => {
+    graph.instance().simulate();
```
Comment (Onatolich) [inline src/components/Graph/Graph.spec.js]: You should not call it here

File: src/components/Graph/Graph.spec.js [hunk 17]
```diff
@@ -0,0 +1,222 @@
...
+    buffer.getLoops = TrafficRunner.getLoops;
+    buffer.tick = TrafficRunner.prototype.tick;
+    TrafficRunner.prototype.tick = jest.fn();
+    TrafficRunner.getLoops = jest.fn();
+    onSimulationEnd = jest.fn();
+    root = {
+      classed: jest.fn(),
+    };
+    graph = shallow(<Graph graph={graphData} onSimulationEnd={onSimulationEnd} />);
+    graph.instance().root = root;
```
Comment (Onatolich) [inline src/components/Graph/Graph.spec.js]: `const instance = graph.instance();` and use it everywhere.

File: src/components/Graph/Graph.spec.js [hunk 18]
```diff
@@ -0,0 +1,222 @@
...
+  expect(graph.hasClass('root')).toBeTruthy();
+});
+
+describe('componentDidMount', () => {
+  let graph;
+  let root;
+  let onSimulationEnd;
+  const buffer = {};
+
+  beforeEach(() => {
```
Comment (Onatolich) [inline src/components/Graph/Graph.spec.js]: Use `jest.spyOn` and `jest.restoreAllMocks`

File: src/components/Graph/Graph.spec.js [hunk 19]
```diff
@@ -0,0 +1,222 @@
...
+    graph.instance().root = root;
+    graph.instance().componentDidMount();
+  });
+
+  afterEach(() => {
+    TrafficRunner.getLoops = buffer.getLoops;
+    TrafficRunner.prototype.tick = buffer.tick;
+  });
+
+  test('should set true to root element ', () => {
```
Comment (Onatolich) [inline src/components/Graph/Graph.spec.js]: Probably it should be `should add simulation class to root element`

File: src/components/Graph/Graph.spec.js [hunk 20]
```diff
@@ -0,0 +1,222 @@
...
+    expect(root.classed).toHaveBeenCalledWith('simulation', true);
+  });
+
+  test('should set false to root element', () => {
```
Comment (Onatolich) [inline src/components/Graph/Graph.spec.js]: Probably it should be `should remove simulation class from root element`

File: src/components/Graph/Graph.spec.js [hunk 21]
```diff
@@ -0,0 +1,222 @@
...
+    graph.instance().simulate();
+    expect(root.classed).toHaveBeenCalledWith('simulation', false);
+  });
+
+  test('should call onSimulationEnd prop with root element', () => {
+    graph.instance().onSimulationEnd();
+    expect(onSimulationEnd).toHaveBeenCalledWith(root);
+  });
+});
```
Comment (Onatolich) [inline src/components/Graph/Graph.spec.js]: Add also tests checking that traffic runner instance is created and set to `trafficRunner` attr, graph nodes are added to nodes collection, and edges - to edges collection.

File: src/components/Graph/Graph.spec.js [hunk 22]
```diff
@@ -0,0 +1,222 @@
+import React from 'react';
```
Comment (Onatolich) [inline src/components/Graph/Graph.spec.js]: Add tests checking that instance have `simulation` attr with instance of Simulation and `nodes` and `edges` attrs with instances of `EntitiesCollection` with corresponded spawn functions. Also methods `onSimulationEnd`, `spawnNode`, `spawnEdge`, `postSimulationCleanup`, `simulate`, `updateSelection` and `applyDiffOnCollection` are not covered by tests at all.

File: src/components/Graph/Graph.spec.js [hunk 23]
```diff
@@ -32,106 +32,145 @@ const graphData = {
...
   let onSimulationEnd;
-  const buffer = {};
+  let instance;
 
   beforeEach(() => {
-    buffer.add = EntitiesCollection.prototype.add;
+    jest.spyOn(TrafficRunner.prototype, 'tick');
+    jest.spyOn(TrafficRunner, 'getLoops');
+    jest.spyOn(EntitiesCollection.prototype, 'add');
+    jest.spyOn(TrafficRunner, 'getLoops');
```
Comment (Onatolich) [inline src/components/Graph/Graph.spec.js]: You already has spy on this method.

File: src/components/Graph/Graph.spec.js [hunk 24]
```diff
@@ -32,106 +32,145 @@ const graphData = {
...
-
-  beforeEach(() => {
-    selectedNode = 1;
-    selectedEdge = { source: 1, target: 2 };
-    graph = shallow(<Graph graph={graphData} />);
-  });
+  const selectedNode = 1;
+  const selectedEdge = { source: 1, target: 2 };
 
   test('should call refresh if new prop graph are not equal to old one', () => {
```
Comment (Onatolich) [inline src/components/Graph/Graph.spec.js]: `should call refresh method if graph prop was changed`

File: src/components/Graph/Graph.spec.js [hunk 25]
```diff
@@ -32,106 +32,145 @@ const graphData = {
...
     });
-    expect(graph.instance().refresh).toHaveBeenCalled();
+
+    expect(instance.refresh).toHaveBeenCalled();
   });
-  test('should call updateSelection if new props selectedNode and selectedEdge are not equal to old ones', () => {
-    graph = shallow(<Graph
-      graph={graphData}
+
+  test('should not call refresh if new prop graph are equal to old one', () => {
```
Comment (Onatolich) [inline src/components/Graph/Graph.spec.js]: `should not call refresh method if graph prop was not changed`

File: src/components/Graph/Graph.spec.js [hunk 26]
```diff
@@ -32,106 +32,145 @@ const graphData = {
...
+    const graphWrapper = shallow(<Graph graph={graph} />);
+    const instance = graphWrapper.instance();
+    instance.refresh = jest.fn();
+    instance.componentWillReceiveProps({ graph });
+    expect(instance.refresh).not.toHaveBeenCalled();
+  });
+
+  test('should call updateSelection if new prop selectedNode is not equal to old ones', () => {
```
Comment (Onatolich) [inline src/components/Graph/Graph.spec.js]: `should call updateSelection method if selectedNode prop was changed`

File: src/components/Graph/Graph.spec.js [hunk 27]
```diff
@@ -32,106 +32,145 @@ const graphData = {
...
-    graph.instance().updateSelection = jest.fn();
-    graph.instance().componentWillReceiveProps({
+
+    const instance = graphWrapper.instance();
+    instance.updateSelection = jest.fn();
+    instance.componentWillReceiveProps({
       selectedNode: 2,
+    });
+
+    expect(instance.updateSelection).toHaveBeenCalled();
```
Comment (Onatolich) [inline src/components/Graph/Graph.spec.js]: toHaveBeenCalledWith

File: src/components/Graph/Graph.spec.js [hunk 28]
```diff
@@ -32,106 +32,145 @@ const graphData = {
...
+    instance.refresh = jest.fn();
+    instance.componentWillReceiveProps({
       graph: {
         nodes: [],
         edges: [],
       },
     });
-    expect(graph.instance().refresh).toHaveBeenCalled();
+
+    expect(instance.refresh).toHaveBeenCalled();
```
Comment (Onatolich) [inline src/components/Graph/Graph.spec.js]: toHaveBeenCalledWith

File: src/components/Graph/Graph.spec.js [hunk 29]
```diff
@@ -32,106 +32,145 @@ const graphData = {
...
+    const instance = graphWrapper.instance();
+    instance.updateSelection = jest.fn();
+    instance.componentWillReceiveProps({
       selectedNode: 2,
+    });
+
+    expect(instance.updateSelection).toHaveBeenCalled();
+  });
+
+  test('should call updateSelection if new prop selectedEdge is not equal to old ones', () => {
```
Comment (Onatolich) [inline src/components/Graph/Graph.spec.js]: `should call updateSelection method if selectedEdge prop was changed`

File: src/components/Graph/Graph.spec.js [hunk 30]
```diff
@@ -32,106 +32,145 @@ const graphData = {
...
+    const graphWrapper = shallow(<Graph
+      graph={graph}
+      selectedEdge={selectedEdge}
+    />);
```
Comment (Onatolich) [inline src/components/Graph/Graph.spec.js]: Use proper multi line formatting everywhere:
```
shallow(
  <Graph
    graph={graph}
    whatever={whatever}
  />
)
```

File: src/components/Graph/Graph.spec.js [hunk 31]
```diff
@@ -32,106 +32,145 @@ const graphData = {
...
+
+    const instance = graphWrapper.instance();
+    instance.updateSelection = jest.fn();
+    instance.componentWillReceiveProps({
       selectedEdge: { source: 2, target: 1 },
     });
-    expect(graph.instance().updateSelection).toHaveBeenCalled();
+
+    expect(instance.updateSelection).toHaveBeenCalled();
```
Comment (Onatolich) [inline src/components/Graph/Graph.spec.js]: toHaveBeenCalledWith

File: src/components/Graph/Graph.spec.js [hunk 32]
```diff
@@ -32,106 +32,145 @@ const graphData = {
...
+  });
+
+  test('should not call updateSelection if new props selectedNode and selectedEdge are equal to old ones', () => {
```
Comment (Onatolich) [inline src/components/Graph/Graph.spec.js]: `should not call updateSelection method if selectedNode and selectedEdge props were not changed`

File: src/components/Graph/Graph.spec.js [hunk 33]
```diff
@@ -140,83 +179,122 @@ test('onEdgeClick should call onEdgeClick prop with edge data', () => {
...
+    nodes: { add: [], modify: [], remove: [] },
+  };
 
   beforeEach(() => {
-    buffer.processGraph = Graph.processGraph;
-    Graph.processGraph = jest.fn().mockReturnValue({ nodes: [], edges: [] });
-    graph = shallow(<Graph graph={graphData} />);
-    instance = graph.instance();
+    jest.spyOn(Graph, 'processGraph');
+    Graph.processGraph = jest.fn().mockReturnValue({ ...refreshData });
```
Comment (Onatolich) [inline src/components/Graph/Graph.spec.js]: This two lines makes no sense. It should be:
`jest.spyOn(Graph, 'processGraph').mockReturnValue({ ...refreshData });`

File: src/components/Graph/Graph.spec.js [hunk 34]
```diff
@@ -140,83 +179,122 @@ test('onEdgeClick should call onEdgeClick prop with edge data', () => {
...
-    nodes: { add: [], modify: [], remove: [] },
-  };
-  const props = {
-    selectedNode: 'selectedNode',
-    selectedEdge: 'selectedEdge',
-    resetSelection: jest.fn(),
-  };
+  test('should call applyDiffOnCollection method for edges', () => {
+    instance.refresh({
+      graph: { nodes: [] },
```
Comment (Onatolich) [inline src/components/Graph/Graph.spec.js]: Probably `nodes` should be `edges` here

File: src/components/Graph/Graph.spec.js [hunk 35]
```diff
@@ -140,83 +179,122 @@ test('onEdgeClick should call onEdgeClick prop with edge data', () => {
...
-    const instance = graph.instance();
-    instance.applyDiffOnSelection(graphDiff);
-    expect(props.resetSelection).not.toHaveBeenCalled();
+    graphWrapper = shallow(<Graph graph={graph} />);
+    instance = graphWrapper.instance();
+    instance.applyDiffOnSelection(diffGraph);
+    expect(resetSelection).not.toHaveBeenCalled();
+  });
+
+  test('should not call resetSelection if selectedNode is not exists', () => {
```
Comment (Onatolich) [inline src/components/Graph/Graph.spec.js]: You're testing only first and last conditions. You should cover next cases:
1. resetSelection should not be called if there is no selectedNode and selectedEdge
2. resetSelection should not be called if there is selected node and it was not removed
3. resetSelection should not be called if there is selected edge and it was not removed
4. resetSelection should be called if there is selected node and it was removed
5. resetSelection should be called if there is selected edge and it was removed

File: src/components/Graph/Graph.spec.js (hunk 30 above)
Comment (itdev) [reply src/components/Graph/Graph.spec.js]: I have formatted it according to eslint rules. Added `function-paren-newline`: off rule

**Pull request closed without merging**